lib/matplotlib/cycler.py

```diff
@@ -0,0 +1,237 @@
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+import six
+from itertools import product, cycle
+from six.moves import zip
+
+
+class Cycler(object):
+    """
+    A class to handle cycling multiple artist properties.
+
+    This class has two compositions methods '+' for 'inner'
+    products of the cycles and '*' for outer products of the
+    cycles.
+
+    Parameters
+    ----------
+    left : Cycler or None
+        The 'left' cycler
+
+    right : Cycler or None
+        The 'right' cycler
+
+    op : func or None
+        Function which composes the 'left' and 'right' cyclers.
+
+    """
+    def __init__(self, left, right=None, op=None):
+        self._left = left
+        self._right = right
+        self._op = op
+        l_key = left.keys if left is not None else set()
+        r_key = right.keys if right is not None else set()
+        if l_key & r_key:
+            raise ValueError("Can not compose overlapping cycles")
+        self._keys = l_key | r_key
+
+    @property
+    def keys(self):
+        return self._keys
+
+    def finite_iter(self):
+        """
+        Return a finite iterator over the configurations in
+        this cycle.
+        """
+        if self._right is None:
+            try:
+                return self._left.finite_iter()
+            except AttributeError:
+                return iter(self._left)
+        return self._compose()
+
+    def _compose(self):
+        """
+        Compose the 'left' and 'right' components of this cycle
+        with the proper operation (zip or product as of now)
+        """
+        for a, b in self._op(self._left.finite_iter(),
+                             self._right.finite_iter()):
+            out = dict()
+            out.update(a)
+            out.update(b)
+            yield out
+
+    @classmethod
+    def from_iter(cls, label, itr):
```

Comment:
A way to make the API nicer would be to define a `cycler` factory function that would call this classmethod, and document that as the standard interface.  It's all a user would need to know about.